### ISSUE TITLE
Snapshot of an empty filesystem still takes up a lot of space in the pool

### ISSUE THREAD

**Issue by agrover** [stratis-storage/stratisd#694]
Shouldn't it take *less* than an entirely separate new fs?

```
[root@fedora-devel test1]# stratis pool list
Name     Total Physical Size  Total Physical Used
test1                 11 GiB           < 1.99 GiB
[root@fedora-devel test1]# stratis fs create test1 fs1
[root@fedora-devel test1]# stratis pool list
Name     Total Physical Size  Total Physical Used
test1                 11 GiB           > 2.52 GiB
[root@fedora-devel test1]# stratis fs snapshot test1 fs1 fs2
[root@fedora-devel test1]# stratis pool list
Name     Total Physical Size  Total Physical Used
test1                 11 GiB           > 3.05 GiB
```

**Comment by mulkieran** [comment]
The snapshot seems to use the same amount of space as its origin.

**Comment by mulkieran** [comment]
The devicemapper code is written that way...the snapshot is always the same size as its origin.

**Comment by agrover** [comment]
Why should it allocate that much? A snapshot should have almost the same data in it as the original -- the only thing that should be different is we give it a new XFS uuid. To a large degree it should just be pointing to existing data blocks rather than allocating new data blocks. My current hunch/uninformed-guess is the change-uuid just happened to modify the same 500MiB of blocks in the original newly-mkfs'ed thindev, causing them all to be CoWed. This seems a little farfetched. To learn more, we could play around with using a smaller data block size maybe? Or different mkfs parameters for XFS? Or maybe ask an XFS expert? @trgill any ideas what might be causing this?

**Comment by mulkieran** [comment]
> A snapshot should have almost the same data in it as the original -- the only thing that should be different is we give it a new XFS uuid. To a large degree it should just be pointing to existing data blocks rather than allocating new data blocks.

That's all true, but the libraries have their reasons of which reason knows nothing. We should at least eliminate devicemapper actions from consideration.

**Comment by mulkieran** [comment]
Devicemapper tests indicate that, at least at moderate sizes, usage is what might be expected: https://github.com/stratis-storage/devicemapper-rs/pull/247.

**Comment by mulkieran** [comment]
In devicemapper tests, creating a filesystem usage was 10240 Sectors, which I believe is exactly 5 MiB. But here, creating a filesystem seems to take up about 1/2 GiB. This, too, is weird.

**Comment by mulkieran** [comment]
We need to break down where that total physical used is actually going.

**Comment by mulkieran** [comment]
So, our devicemapper tests seem to confirm original hypothesis. A 1 TiB filesystem really does seem to use up about 1/2 GiB with its metadata. Taking a snapshot requires 0 additional usage, but updating the UUID of the filesystem causes pool usage roughly the same, although a little bit less in tests, than the usage of the origin: https://github.com/stratis-storage/devicemapper-rs/pull/247.

**Comment by agrover** [comment]
We've determined this is because the entire XFS log (~500MiB) is cleared when a new UUID is given.

Changing this behavior is an XFS issue. Moving milestone to Future.

**Comment by drckeefe** [comment]
Adding a thought to this issue for later consideration: We might consider an option to not give a new UUID.  If the user is interested in snapshots for point in time versions of the filesytem, then they might not care to have the UUID changed.  This should save space.  Of course, there would be some limitations to the snapshots that don't have an updated UUID, such as, they are not mountable unless the UUID is changed.  If a user wants to use the snapshot as a new functioning filesystem then the UUID could be changed and they should be fine with the 500MiB cost.
